services-api/src/main/java/io/scalecube/services/Reflect.java

```diff
@@ -62,23 +65,48 @@ public static Type parameterizedReturnType(Method method) {
       if (method.isAnnotationPresent(RequestType.class)) {
         return method.getAnnotation(RequestType.class).value();
       } else {
-        if (method.getGenericParameterTypes()[0] instanceof ParameterizedType) {
+        Integer requestParameterIndex = requestParameterIndex(method);
+
+        if (requestParameterIndex == null) {
+          return Void.TYPE;
+        }
+
+        if (method.getGenericParameterTypes()[requestParameterIndex] instanceof ParameterizedType) {
           try {
             return Class.forName(parameterizedRequestType(method).getTypeName());
           } catch (ClassNotFoundException e) {
             return Object.class;
           }
-        } else if (ServiceMessage.class.equals(method.getParameterTypes()[0])) {
+        } else if (ServiceMessage.class.equals(method.getParameterTypes()[requestParameterIndex])) {
           return Object.class;
         } else {
-          return method.getParameterTypes()[0];
+          return method.getParameterTypes()[requestParameterIndex];
         }
       }
     } else {
       return Void.TYPE;
     }
   }
 
+  /**
+   * Return index of parameter which is defined as request.
+   *
+   * @param method method
+   * @return index of parameter
+   */
+  public static Integer requestParameterIndex(Method method) {
```

Comment:
`OptionalInt` can be nicer

Reply:
I changed implementation, so `requestParameterIndex` does not exist now.